sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,320 @@
+        this.elemCoder = elemCoder;
+      }
+
+      @Override
+      public void encode(Checkpoint<T> value, OutputStream outStream, Context context)
+          throws CoderException, IOException {
+        elemsCoder.encode(value.residualElements, outStream, context);
+        serializableCoder.encode(value.residualSource, outStream, context);
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public Checkpoint<T> decode(InputStream inStream, Context context)
+          throws CoderException, IOException {
+        return new Checkpoint<>(
+            elemsCoder.decode(inStream, context),
+            serializableCoder.decode(inStream, context));
+      }
+
+      @Override
+      public List<Coder<?>> getCoderArguments() {
+        return Arrays.<Coder<?>>asList(elemCoder);
+      }
+    }
+
+    @VisibleForTesting
+    class Reader extends UnboundedReader<T> {
+      private final PipelineOptions options;
+
+      private final List<TimestampedValue<T>> residualElements;
+      private @Nullable BoundedSource<T> residualSource;
+
+      private boolean currentElementInList;
+      private int currentIndexInList;
+      private BoundedReader<T> reader;
+      private boolean done;
+
+      public Reader(
+          List<TimestampedValue<T>> residualElements,
+          @Nullable BoundedSource<T> residualSource,
+          PipelineOptions options) {
+        this.residualElements = Preconditions.checkNotNull(residualElements, "residualElements");
+        this.residualSource = residualSource;
+        this.options = Preconditions.checkNotNull(options, "options");
+        this.currentElementInList = true;
+        this.currentIndexInList = -1;
+        this.reader = null;
+        this.done = false;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        return advance();
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        if (advanceInList()) {
+          done = false;
+        } else {
+          done = !advanceInReader();
+        }
+        return !done;
+      }
+
+      private boolean advanceInList() {
+        if (residualElements != null && currentIndexInList + 1 < residualElements.size()) {
+          ++currentIndexInList;
+          currentElementInList = true;
+          return true;
+        } else {
+          currentElementInList = false;
+          return false;
+        }
+      }
+
+      private boolean advanceInReader() throws IOException {
+        boolean hasNext;
+        if (boundedSource == null) {
+          hasNext = false;
+        } else if (reader == null) {
+          reader = residualSource.createReader(options);
+          hasNext = reader.start();
+        } else {
+          hasNext = reader.advance();
+        }
+        return hasNext;
+      }
+
+      @Override
+      public void close() throws IOException {
+        reader.close();
```

Comment:
It's probably better to close the underlying BoundedReader once it is out of elements.  Since that is a permanent state for a BoundedReader, there's no need to wait until the enclosing UnboundedReader is closed.